src/app/modules/form/admin-form/admin-form.service.ts

```diff
@@ -410,3 +418,83 @@ export const duplicateForm = (
     },
   )
 }
+
+/**
+ * Updates form fields of given form depending on the given editFormFieldParams
+ * @param originalForm the original form to update form fields for
+ * @param editFormFieldParams the parameters stating the type of updates and the position metadata if update type is edit
+ *
+ * @returns ok(updated form) if form fields update successfully
+ * @returns err(EditFieldError) if invalid updates are being performed to form fields
+ * @returns err(set of DatabaseError) if any database errors occurs
+ */
+export const editFormFields = (
+  originalForm: IPopulatedForm,
+  editFormFieldParams: EditFormFieldParams,
+): ResultAsync<
+  IPopulatedForm,
+  EditFieldError | ReturnType<typeof transformMongoError>
+> => {
+  // TODO(#815): Split out this function into their own separate service functions depending on the update type.
+  return getUpdatedFormFields(
+    originalForm.form_fields,
+    editFormFieldParams,
+  ).asyncAndThen((newFormFields) => {
+    // Update form fields of original form.
+    originalForm.form_fields = newFormFields
+    return ResultAsync.fromPromise(originalForm.save(), (error) => {
+      logger.error({
+        message: 'Error encountered while editing form fields',
+        meta: {
+          action: 'editFormFields',
+          originalForm,
+          editFormFieldParams,
+        },
+        error,
+      })
+
+      return transformMongoError(error)
+    })
+  })
+}
+
+/**
+ * Updates form schema of given form depending on the given formUpdateParams
+ * @param originalForm the original form to update form fields for
+ * @param formUpdateParams the parameters to update the original form with
+ *
+ * @returns ok(updated form) if form fields update successfully
+ * @returns err(set of DatabaseError) if any database errors occurs
+ */
+export const updateForm = (
+  originalForm: IPopulatedForm,
+  formUpdateParams: Except<FormUpdateParams, 'editFormField'>,
```

Comment:
I'm not sure I understand the typing here. `FormUpdateParams` only has an optional `editFormField` key, so if you exclude it then don't you end up with an empty object? this is what I get when I hover over:

<img width="725" alt="Screenshot 2020-12-28 at 10 21 04 AM" src="https://user-images.githubusercontent.com/29480346/103185225-65159500-48f6-11eb-8b58-a1bfa3cf4bb9.png">

Reply:
Good catch, fixed the typing with all the types synced to our Joi validator